src/Microsoft.Azure.ServiceBus/Core/Batch.cs

```diff
@@ -0,0 +1,144 @@
+        public async Task<bool> TryAdd(Message message)
+        {
+            ThrowIfDisposed();
+
+            message.VerifyMessageIsNotPreviouslyReceived();
+
+            var processedMessage = await pluginsCallback(message);
+
+            var amqpMessage = AmqpMessageConverter.SBMessageToAmqpMessage(processedMessage);
+
+            if (firstMessage == null)
+            {
+                originalMessageData = (processedMessage.MessageId, processedMessage.SessionId, processedMessage.PartitionKey, processedMessage.ViaPartitionKey);
+                firstMessage = amqpMessage;
+            }
+
+            var data = AmqpMessageConverter.ToData(amqpMessage);
+            datas.Add(data);
+
+            if (Size <= maximumBatchSize)
+            {
+                return true;
+            }
+
+            datas.Remove(data);
+            return false;
+
+        }
+
+        /// <summary>
+        /// Number of messages in batch.
+        /// </summary>
+        public int Length => datas.Count;
+
+        internal ulong Size => (ulong) result.SerializedMessageSize;
+
+
+        /// <summary>
+        /// Convert batch to AMQP message.
+        /// </summary>
+        /// <returns></returns>
+        public AmqpMessage ToAmqpMessage()
+        {
+            ThrowIfDisposed();
+
+            if (datas.Count == 1)
+            {
+                firstMessage.Batchable = true;
+                return firstMessage;
+            }
+
+            if (originalMessageData.messageId != null)
+            {
+                result.Properties.MessageId = originalMessageData.messageId;
+            }
+
+            if (originalMessageData.sessionId != null)
+            {
+                result.Properties.GroupId = originalMessageData.sessionId;
+            }
+
+            if (originalMessageData.partitionKey != null)
+            {
+                result.MessageAnnotations.Map[AmqpMessageConverter.PartitionKeyName] = originalMessageData.partitionKey;
+            }
+
+            if (originalMessageData.viaPartitionKey != null)
+            {
+                result.MessageAnnotations.Map[AmqpMessageConverter.ViaPartitionKeyName] = originalMessageData.viaPartitionKey;
+            }
+
+            result.MessageFormat = AmqpConstants.AmqpBatchedMessageFormat;
+            result.Batchable = true;
+            return result;
+        }
+
+        public void Dispose()
+        {
+            // TODO: review if there's anything else to do
+            firstMessage?.Dispose();
+            result?.Dispose();
+
+            firstMessage = null;
+            result = null;
+        }
+
+        private void ThrowIfDisposed()
+        {
+            if (result == null)
+            {
+                throw new Exception("Batch is has been disposed and cannot be re-used.");
+            }
+        }
+
+        private string DebuggerDisplay => $"Batch: size={Size} message count={datas.Count} maximum size={maximumBatchSize}";
```

Comment:
>size={Size} message count= [](start = 51, length = 26)

Add a `.` or `;` or any identifier between two properties so that its easier to read